### PR TITLE
Support varying spatial resoln along the spectral axis

### PR DESCRIPTION
**Pull request by keflavich**
Infrastructure for handling varying resolution from new CASA tclean cubes

cc @low-sky, @e-koch, @astrofrog

See related https://github.com/radio-astro-tools/radio_beam/pull/35

### CODE REVIEW

**Comment by e-koch** [comment]
:+1: for multi-beam tests

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -2287,6 +2290,203 @@ def find_lines(self, velocity_offset=None, velocity_convention=None,
 
         return result
 
+    def reproject(self, header, order='bilinear'):
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py]: `@warn_slow` or a warning for potentially needing a lot of memory

Comment (keflavich) [inline spectral_cube/spectral_cube.py]: Yep, but this belongs on https://github.com/radio-astro-tools/spectral-cube/pull/289

**Comment by keflavich** [comment]
Yay, all passing!  I think this is ready to merge, though since it's big I'd like a review from @astrofrog or @e-koch.  I might wait a few days at least and do some local testing (real use) to see how it fares.

**Comment by keflavich** [comment]
I think we now have an interdependency issue for the tests between this and `radio_beam`

**Comment by keflavich** [comment]
The latest error has something to do with travis reconfiguration; apparently the wrong numpy version is being run or something like that?

**Comment by keflavich** [comment]
One of the failures was that conda doesn't have bottleneck for python3.5, only python3.4.  I don't know why...

**Comment by keflavich** [comment]
yay, improved coverage!  But, I can't reproduce these numpy 1.9 errors locally.

File: spectral_cube/spectral_cube.py [hunk 2]
```diff
@@ -2287,6 +2299,363 @@ def find_lines(self, velocity_offset=None, velocity_convention=None,
...
+            if maxdiff > th:
+                errormessage += ("Beam {2}s differ by up to {0}x, which is greater"
+                                 " than the threshold {1}\n".format(maxdiff,
+                                                                    threshold,
+                                                                    qtyname
+                                                                   ))
+        if errormessage != "":
+            raise ValueError(errormessage)
+
+    def _average_beams(self, threshold):
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py]: `threshold` isn't being checked in here. The `_check_beam_areas` in `_handle_beam_areas_wrapper` could be moved into here to ensure the threshold check is performed for any beam averaging.

File: spectral_cube/spectral_cube.py [hunk 3]
```diff
@@ -2287,6 +2299,363 @@ def find_lines(self, velocity_offset=None, velocity_convention=None,
...
+            # check that the spectral axis is being operated over
+            # moments are a special case b/c they default to axis=0
+            need_to_handle_beams = ('axis' in kwargs and kwargs['axis']==0
+                                    or (hasattr(kwargs['axis'], '__len__') and
+                                        0 in kwargs['axis'])
+                                    or ('axis' not in kwargs and 'moment' in
+                                        function.__name__))
+
+            if need_to_handle_beams:
+                self._check_beam_areas(beam_threshold)
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py]: See comment above.

File: spectral_cube/spectral_cube.py [hunk 4]
```diff
@@ -2287,6 +2299,363 @@ def find_lines(self, velocity_offset=None, velocity_convention=None,
...
+
+        qtys = dict(area = u.Quantity(self.beams, u.sr),
+                    major = u.Quantity([bm.major for bm in self.beams], u.deg),
+                    minor = u.Quantity([bm.minor for bm in self.beams], u.deg),
+                    pa = u.Quantity([bm.pa for bm in self.beams], u.deg),
+                   )
+
+        errormessage = ""
+
+        for qtyname, qty in qtys.items():
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py]: `cube_utils.average_beams` should be used to get the mean values (for consistency).

For the PA, I guess the comparison should be with the max absolute difference from the mean?

File: spectral_cube/cube_utils.py [hunk 5]
```diff
@@ -230,3 +232,42 @@ def iterator_strategy(cube, axis=None):
...
+    """
+
+    from radio_beam import Beam
+    from astropy.stats import circmean
+
+    major = u.Quantity([bm.major for bm in beams], u.deg)
+    minor = u.Quantity([bm.minor for bm in beams], u.deg)
+    pa = u.Quantity([bm.pa for bm in beams], u.deg)
+    new_beam = Beam(major=major.mean(), minor=minor.mean(),
+                    pa=circmean(pa, weights=major/minor))
```
Comment (e-koch) [inline spectral_cube/cube_utils.py]: Is there a big difference if you just weight be `major`? The minor axis should be independent from the vector defining the PA. `major/minor` weights the most asymmetric beams the most, and I'm not sure if that's what we want. I guess the most asymmetric beams are where the PA matters the most.

**Comment by e-koch** [comment]
@keflavich - I've added some minor comments. I haven't had a chance to test locally yet (though I'm working on the PR branch and it hasn't broken my normal workflow).

File: spectral_cube/cube_utils.py (hunk 5 above)
Comment (keflavich) [inline spectral_cube/cube_utils.py]: That was my thinking.  I don't know what a correct weighting scheme would be; probably needs to be determined

**Comment by keflavich** [comment]
OK, if this passes I'll merge.  But changelog entry, right...

**Comment by e-koch** [comment]
:+1:

**Pull request merged**